File: internal/opamrepo/opamrepo.go
// Package opamrepo locates packages inside an opam-repository checkout.
package opamrepo

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
)

// Package is one versioned package in a repository.
type Package struct {
	Name    string // "posix-base"
	Version string // "2.0.0"
	Dir     string // <root>/packages/posix-base/posix-base.2.0.0
}

// OpamPath is the path of the package's opam file.
func (p Package) OpamPath() string { return filepath.Join(p.Dir, "opam") }

// RelDir is the package directory relative to the repository root.
func (p Package) RelDir() string {
	return filepath.Join("packages", p.Name, p.Name+"."+p.Version)
}

// Sources is the set of repositories a mirror is built from.
//
// Overlays are searched before Base. This is what makes opam-monorepo usable
// through a mirror: dune-universe/opam-overlays redefines a number of packages
// so that they build with dune, and those definitions have to win over the
// official ones. Merging repositories is a thing you can do when you hold the
// directory yourself, and cannot when you only proxy it.
type Sources struct {
	Base     string
	Overlays []string
}

// Roots returns the search path, highest priority first.
func (s Sources) Roots() []string {
	roots := make([]string, 0, len(s.Overlays)+1)
	roots = append(roots, s.Overlays...)
	if s.Base != "" {
		roots = append(roots, s.Base)
	}
	return roots
}

// List returns every versioned package visible across roots.
//
// Roots are searched in order and the first one to provide a given name and
// version wins, so a later root never shadows an earlier one.
//
// The version comes from the directory name, which is where opam itself reads
// it from. Asking an external tool for it, as tempting as it looks, is both
// slower and less accurate: the directory name is the authority.
func List(roots []string) ([]Package, error) {
	var out []Package
	seen := make(map[string]bool)

	for _, root := range roots {
		packages := filepath.Join(root, "packages")
		entries, err := os.ReadDir(packages)
		if err != nil {
			return nil, fmt.Errorf("reading %s: %w", packages, err)
		}

		for _, nameDir := range entries {
			if !nameDir.IsDir() {
				continue
			}
			name := nameDir.Name()

			versions, err := os.ReadDir(filepath.Join(packages, name))
			if err != nil {
				return nil, err
			}
			for _, versionDir := range versions {
				if !versionDir.IsDir() {
					continue
				}
				base := versionDir.Name()
				version, ok := versionOf(name, base)
				if !ok {
					continue
				}
				if seen[base] {
					continue // provided by a higher priority root
				}
				dir := filepath.Join(packages, name, base)
				if _, err := os.Stat(filepath.Join(dir, "opam")); err != nil {
					continue // a directory without an opam file is not a package
				}
				seen[base] = true
				out = append(out, Package{Name: name, Version: version, Dir: dir})
			}
		}
	}
	return out, nil
}

// versionOf splits "posix-base.2.0.0" into its version part.
func versionOf(name, dirName string) (string, bool) {
	rest, ok := strings.CutPrefix(dirName, name+".")
	if !ok || rest == "" {
		return "", false
	}
	return rest, true
}

// Find resolves one package by name and version, searching roots in order.
func Find(roots []string, name, version string) (Package, error) {
	if err := ValidName(name); err != nil {
		return Package{}, err
	}
	if err := ValidVersion(version); err != nil {
		return Package{}, err
	}
	for _, root := range roots {
		dir := filepath.Join(root, "packages", name, name+"."+version)
		if _, err := os.Stat(filepath.Join(dir, "opam")); err == nil {
			return Package{Name: name, Version: version, Dir: dir}, nil
		}
	}
	return Package{}, fs.ErrNotExist
}

// ValidName rejects anything that could escape the packages directory.
//
// These names arrive from HTTP request paths, so this is the boundary between
// a request and the filesystem. Everything downstream assumes it ran.
func ValidName(s string) error { return validPathElement("package name", s) }

// ValidVersion applies the same rule to a version string.
func ValidVersion(s string) error { return validPathElement("version", s) }

func validPathElement(what, s string) error {
	if s == "" {
		return fmt.Errorf("empty %s", what)
	}
	if len(s) > 128 {
		return fmt.Errorf("%s too long", what)
	}
	if s == "." || s == ".." || strings.Contains(s, "/") || strings.Contains(s, `\`) {
		return fmt.Errorf("invalid %s %q", what, s)
	}
	for _, r := range s {
		switch {
		case r >= 'a' && r <= 'z', r >= 'A' && r <= 'Z', r >= '0' && r <= '9':
		case r == '.', r == '-', r == '_', r == '+', r == '~':
		default:
			return fmt.Errorf("invalid character %q in %s %q", r, what, s)
		}
	}
	return nil
}

File: ocaml/lib/opamrepo.ml
(* Locate packages inside one or more opam-repository checkouts. Port of the Go
   package: the version comes from the directory name, which is where opam reads
   it from, and overlays are searched before the base repository. *)

type package = { name : string; version : string; dir : string }

let opam_path (p : package) = Filename.concat p.dir "opam"

let rel_dir (p : package) =
  Filename.concat (Filename.concat "packages" p.name) (p.name ^ "." ^ p.version)

(* The set of repositories a mirror is built from. Overlays win over base, which
   is what makes opam-monorepo usable through a mirror: dune-universe overlays
   redefine packages to build with dune, and those must take precedence. *)
type sources = { base : string; overlays : string list }

let roots (s : sources) : string list =
  s.overlays @ (if s.base = "" then [] else [ s.base ])

let valid_element what s =
  if s = "" then Error (Printf.sprintf "empty %s" what)
  else if String.length s > 128 then Error (what ^ " too long")
  else if s = "." || s = ".." || String.contains s '/' || String.contains s '\\'
  then Error (Printf.sprintf "invalid %s %S" what s)
  else begin
    let ok = ref true in
    String.iter
      (fun c ->
        match c with
        | 'a' .. 'z' | 'A' .. 'Z' | '0' .. '9' | '.' | '-' | '_' | '+' | '~' -> ()
        | _ -> ok := false)
      s;
    if !ok then Ok () else Error (Printf.sprintf "invalid character in %s %S" what s)
  end

let valid_name = valid_element "package name"
let valid_version = valid_element "version"

let version_of name dirname =
  let prefix = name ^ "." in
  let pl = String.length prefix in
  if String.length dirname > pl && String.sub dirname 0 pl = prefix then
    Some (String.sub dirname pl (String.length dirname - pl))
  else None

let is_dir p = try Sys.is_directory p with Sys_error _ -> false
let has_opam dir = Sys.file_exists (Filename.concat dir "opam")

(* Every versioned package visible across roots. Roots are searched in order and
   the first to provide a given name.version wins, so a later root never shadows
   an earlier one. *)
let list (roots : string list) : package list =
  let seen = Hashtbl.create 4096 in
  let acc = ref [] in
  List.iter
    (fun root ->
      let packages = Filename.concat root "packages" in
      if is_dir packages then
        Array.iter
          (fun name ->
            let namedir = Filename.concat packages name in
            if is_dir namedir then
              Array.iter
                (fun base ->
                  match version_of name base with
                  | None -> ()
                  | Some version ->
                      if not (Hashtbl.mem seen base) then begin
                        let dir = Filename.concat namedir base in
                        if has_opam dir then begin
                          Hashtbl.add seen base ();
                          acc := { name; version; dir } :: !acc
                        end
                      end)
                (Sys.readdir namedir))
          (Sys.readdir packages))
    roots;
  List.rev !acc

(* Resolve one package by name and version, searching roots in order. These
   values arrive from HTTP request paths, so this is the boundary between a
   request and the filesystem. *)
let find (roots : string list) name version : (package, string) result =
  match valid_name name with
  | Error e -> Error e
  | Ok () -> (
      match valid_version version with
      | Error e -> Error e
      | Ok () ->
          let base = name ^ "." ^ version in
          let rec go = function
            | [] -> Error "not found"
            | root :: rest ->
                let dir =
                  Filename.concat
                    (Filename.concat (Filename.concat root "packages") name)
                    base
                in
                if has_opam dir then Ok { name; version; dir } else go rest
          in
          go roots)
